pkg/cmd/cli/cmd/loginoptions.go

```diff
@@ -0,0 +1,325 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/golang/glog"
+	"github.com/openshift/origin/pkg/client"
+
+	kerrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	kclientcmd "github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
+	clientcmdapi "github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd/api"
+	"github.com/openshift/origin/pkg/cmd/cli/config"
+	"github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	osclientcmd "github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/cmd/util/tokencmd"
+	"github.com/openshift/origin/pkg/user/api"
+)
+
+const defaultClusterURL = "https://localhost:8443"
+
+// Helper for the login and setup process, gathers all information required for a
+// successful login and eventual update of config files.
+// Depending on the Reader present it can be interactive, asking for terminal input in
+// case of any missing information.
+type LoginOptions struct {
+	// flags and printing helpers
+	Username string
+	Password string
+	Project  string
+
+	// infra
+	ClientConfig clientcmd.ClientConfig
+	Config       *kclient.Config
+	Reader       io.Reader
+
+	// flow controllers
+	gatherServerInfo  bool
+	gatherAuthInfo    bool
+	gatherProjectInfo bool
+}
+
+// Gather all required information in a comprehensive order.
+func (o *LoginOptions) GatherInfo() error {
+	if err := o.GatherServerInfo(); err != nil {
+		return err
+	}
+	if err := o.GatherAuthInfo(); err != nil {
+		return err
+	}
+	if err := o.GatherProjectInfo(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Makes sure it has all the needed information about the server we are connecting to,
+// particularly the host address and certificate information. For every information not
+// present ask for interactive user input. Will also ping the server to make sure we can
+// connect to it, and if any problem is found (e.g. certificate issues), ask the user about
+// connecting insecurely.
+func (o *LoginOptions) GatherServerInfo() error {
+	// we need to have a server to talk to
+
+	if util.IsTerminal(o.Reader) {
+		for !o.serverProvided() {
```

Comment:
TODO, always prompt with if they didn't specify --server on the command line.  That  way it's easy to re-login as a different user to the same server and easy to switch servers via the same command and args as your initial login